diff --git a/st.c b/st.c
index abbbe4b..6b4b6e6 100644
--- a/st.c
+++ b/st.c
@@ -1890,10 +1890,8 @@ strhandle(void)
 				fprintf(stderr, "erresc: invalid color j=%d, p=%s\n",
 				        j, p ? p : "(null)");
 			} else {
-				/*
-				 * TODO if defaultbg color is changed, borders
-				 * are dirty
-				 */
+				if (j == defaultbg)
+					xclearwin();
 				redraw();
 			}
 			return;
diff --git a/win.h b/win.h
index e6e4369..323dfae 100644
--- a/win.h
+++ b/win.h
@@ -38,3 +38,4 @@ void xsetpointermotion(int);
 void xsetsel(char *);
 int xstartdraw(void);
 void xximspot(int, int);
+void xclearwin(void);
diff --git a/x.c b/x.c
index 120e495..ebd2c78 100644
--- a/x.c
+++ b/x.c
@@ -825,6 +825,12 @@ xclear(int x1, int y1, int x2, int y2)
 			x1, y1, x2-x1, y2-y1);
 }
 
+void
+xclearwin(void)
+{
+	xclear(0, 0, win.w, win.h);
+}
+
 void
 xhints(void)
 {
